---
title: ext/fig 5
subtitle:  "Running GSEA w/ fGSEA"
author: "pz"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    self_contained: true
    highlight: tango
    df_print: paged
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
    keep_md: yes
    fig_caption: true
    code_folding: hide
  html_notebook:
    self_contained: true
    highlight: tango
    df_print: paged
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, result='hold',fig.width=10,tidy=TRUE, cache=FALSE)
knitr::opts_knit$set(progress=TRUE,verbose=TRUE)
```

```{r libraries, cache=FALSE}
library(data.table)
library(magrittr)
library(knitr)
library(ggplot2); theme_set(theme_bw(base_size = 16))
library(DESeq2)
library(dplyr)
library(pheatmap)
library(RColorBrewer)
library(fgsea)
library(openxlsx)
library(AnnotationDbi)
library(data.table)
library(clusterProfiler)
library(psych)
rgb2hex <- function(r,g,b) sprintf('#%s',paste(as.hexmode(c(r,g,b)),collapse = ''))
```



```{r decoder}
decoderFile <- "decoder.txt"
decoder.data <- fread(decoderFile) %>% as.data.frame()
decoder.data$group <- factor(decoder.data$group, levels = c("N", "E6", "E12", "E24", "L6", "L12", "L24"))
decoder.data$type <- factor(decoder.data$type, levels = c("naive", "effector", "tumor"))
decoder.data$time <- factor(decoder.data$time, levels = c("0hr", "6hr", "12hr", "24hr"))
decoder.data$replicate <- factor(decoder.data$replicate)
decoder.data = decoder.data[order(decoder.data$group),]
row.names(decoder.data) = decoder.data$sample.ID
```

```{r counts}
counts <- read.table(file = "gene.counts.txt", header = TRUE, check.names=FALSE, row.names=1)
decoder.data <- decoder.data[decoder.data$sample.ID %in% colnames(counts),]

if (!identical(colnames(counts), decoder.data$sample.ID)) {
  stop("Column names in counts do not match sample IDs in decoder data")
}

counts <- counts[,c(decoder.data$sample.ID)]
```

# Samples

The following samples were part of this analysis:

```{r samples}
as_tibble(decoder.data)
```


# Differential expression analysis

```{r createDDS}
dds <- DESeq2::DESeqDataSetFromMatrix(countData = counts, 
                                      colData = decoder.data, 
                                      design = ~ group)

keep <- edgeR::filterByExpr(counts(dds), design = model.matrix(~ group, data = colData(dds)))
dds <- dds[keep,]

dds <- estimateSizeFactors(dds)
vsd <- DESeq2::vst(dds, blind = TRUE)
normCounts = assay(vsd)
dds <- DESeq(dds)
```

```{r deseq2}
varInt="group"
de.all <- list()
alphaTR = 0.05
lfcThres = 0

for (comp in combn(nlevels(colData(dds)[,varInt]), 2, simplify=FALSE)){
  levelRef <- levels(colData(dds)[,varInt])[comp[1]]
  levelTest <- levels(colData(dds)[,varInt])[comp[2]]
  de.all[[paste0(levelTest,"_vs_",levelRef)]] <- 
    results(dds, contrast=c(varInt, levelTest, levelRef), alpha=alphaTR) %>% as.data.frame()
}
```

The following genes were detected as differentially expressed (adjusted p < `r alphaTR` and abs(log2FoldChange) > `r lfcThres`):

```{r printDEGtable}
de.all.sig <- lapply(de.all, subset, padj < alphaTR & abs(log2FoldChange) > lfcThres)

res =  as.data.frame(sapply(de.all.sig, nrow))
colnames(res) <- paste0("padj<",alphaTR, " & ","abs(log2FoldChange)>",lfcThres)

rel.contrs = 1:nrow(res)

as_tibble(cbind(res,"contrast #"=rel.contrs) %>%  tibble::rownames_to_column(var="contrast"))
```


# Gene set enrichment with fGSEA

We used the fgseaMultilevel function from the fgsea package with the following paramters were used: 

  -minGSSize = 15
  -maxGSSize  = 500.

Genes were sorted on DESeq2 Wald stat.

Gene sets were either custom curated or accessed from MSigDB using the `msigdbr::msigdbr` function.


```{r loadGeneSets}
gmt.h <- msigdbr::msigdbr(species = "Mus musculus", category = "H") %>%  dplyr::select(gs_name, entrez_gene)  %>% split(x = .$entrez_gene, f = .$gs_name)  
gmt.kegg <- msigdbr::msigdbr(species = "Mus musculus", category = "C2", subcategory = "CP:KEGG") %>% dplyr::select(gs_name, entrez_gene)  %>% split(x = .$entrez_gene, f = .$gs_name)
gmt.reactome <- msigdbr::msigdbr(species = "Mus musculus", category = "C2", subcategory = "CP:REACTOME") %>% dplyr::select(gs_name, entrez_gene)  %>% split(x = .$entrez_gene, f = .$gs_name)
gmt.immunesigdb <- msigdbr::msigdbr(species = "Mus musculus", category = "C7", subcategory = "IMMUNESIGDB") %>% dplyr::select(gs_name, entrez_gene)  %>% split(x = .$entrez_gene, f = .$gs_name)

gmt.custom = gmtPathways("Dysfunction_Exhaustion_Tolerance_Effector_Gene_Sets_ens.gmt")
```


```{r fgsea}
dir.create("./rnaseq_fgsea_out/", showWarnings = FALSE, recursive = FALSE, mode = "0777")

lapply(rel.contrs, function(x){
  name = names(de.all.sig)[x]
  res = de.all[[x]] %>%
    as.data.frame() %>%
    filter(is.finite(stat)) %>%
    as.data.table(keep.rownames=TRUE) %>%
    setnames("rn", "gene_symbol") %>%
    setorder(-stat)
  
  eg = bitr(res$gene_symbol, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db") %>%
    as.data.table() %>%
    setnames(names(.), c("gene_symbol", "entrez"))
  
  eg = res[eg, on = "gene_symbol"] %>%
    setorder(-stat)
  
  geneList = setNames(eg$stat, eg$entrez)
  
  gmtList = list(gmt.h=gmt.h, gmt.kegg=gmt.kegg, gmt.reactome=gmt.reactome, gmt.immunesigdb=gmt.immunesigdb, gmt.custom=gmt.custom)
  
  fgsea.gmt.list = lapply(gmtList, function(gmt){
    set.seed(08123815)
    fgseaRes = fgseaMultilevel(gmt, geneList, minSize=15, maxSize=500)
    fgseaRes = fgseaRes[padj < 0.05]
    fgseaRes$leadingEdge = lapply(fgseaRes$leadingEdge, mapIds, x=org.Mm.eg.db, keytype="ENTREZID", column="SYMBOL")
    return(fgseaRes)
  })
  
  names(fgsea.gmt.list) = c("hallmark", "kegg", "reactome", "immunesigdb", "custom")
  
  openxlsx::write.xlsx(fgsea.gmt.list, paste0("./rnaseq_fgsea_out/deseq2_",name,"_fgsea","_padj0.05.res.xlsx"))
  
})  %>% invisible()
```


# Session Info
```{r session, message=FALSE, warning=FALSE, cache=FALSE,echo=FALSE, fig.width=10, fig.height=5.5, context="data"}
sessionInfo()
```