---
title: Ext Fig 7
subtitle:  "Chromatin accessibility heatmap."
author: "pz"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    highlight: tango
    df_print: paged
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
    code_folding: hide
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, result='hold',fig.width=10,tidy=TRUE, cache=FALSE)
knitr::opts_knit$set(progress=TRUE,verbose=TRUE)
```

```{r libraries, cache=FALSE}
library(data.table)
library(magrittr)
library(knitr)
library(ggplot2); theme_set(theme_bw(base_size = 16))
library(DESeq2)
library(tidyverse)
library(eulerr)
library(profileplyr)
library(grid)
library(ComplexHeatmap)
library(RColorBrewer)
```



```{r decoder}
decoderFile <- "atac_decoder.csv"
decoder.data <- read.table(decoderFile,header=T,stringsAsFactors=F, sep=",")
decoder.data = subset(decoder.data, !(Condition %in% c("P10d", "P1d", "P5d")))
decoder.data = subset(decoder.data, !(type %in% c("lesion")))
decoder.data$time <- factor(decoder.data$time, levels = c("0hr", "6hr", "12hr", "24hr", "5d", "7d", "0d"))
decoder.data$SampleID = make.names(decoder.data$SampleID)
decoder.data$Condition = factor(decoder.data$Condition)
decoder.data$Condition <- relevel(decoder.data$Condition, ref = "N")
decoder.data = with(decoder.data, decoder.data[order(time),])
```


```{r loadDESeq2Object}
# see runDiffBind.R, runSVA.R
load("sva_dds_vsd.RData")
normCounts = assay(vsd)
```

# Samples

The following samples were part of this analysis:

```{r samples}
as_tibble(decoder.data)
```

# Differential accessibility analysis


```{r da}
varInt="Condition"
dbs.all <- list()
alphaTR = 0.05
lfcThres = 1.5

comps = c("N_vs_E6hr", "E6hr_vs_E12hr", "E12hr_vs_E24hr", "E24hr_vs_E5d", "E5d_vs_E7d", "E7d_vs_M")
for (comp in comps){
  levelTest <- strsplit(comp, "_vs_")[[1]][1]
  levelRef <- strsplit(comp, "_vs_")[[1]][2]
  dbs.all[[paste0(levelTest,"_vs_",levelRef)]] <- 
    results(dds, contrast=c(varInt, levelTest, levelRef), alpha=alphaTR) %>% as.data.frame()
}

names(dbs.all) = gsub("hr", "", names(dbs.all))
```


The following genes were detected as differentially expressed (adjusted p < `r alphaTR` and abs(log2FoldChange) > `r lfcThres`):

```{r printDEGtable}
dbs.all.sig <- lapply(names(dbs.all), function(x){
  subset(dbs.all[[x]], padj < alphaTR & abs(log2FoldChange) > lfcThres)
   })
names(dbs.all.sig) <- names(dbs.all)

res = data.frame(sapply(dbs.all.sig, NROW))
colnames(res) <-  paste0("padj<",alphaTR, " & ","abs(log2FoldChange)>",lfcThres)

rel.contrs = 1:nrow(res)

as_tibble(cbind(res,"contrast #"=rel.contrs) %>%  tibble::rownames_to_column(var="contrast"))
```

# Heatmap infection (k = 6)


```{r}
all.sig.rows <- unique(unlist(lapply(dbs.all.sig, row.names)))

annot_col <- data.frame(cond = decoder.data$Condition, row.names = make.names(decoder.data$SampleID))
log2.count.matrix <- normCounts[all.sig.rows, make.names(decoder.data$SampleID)]

#Obtain the z-score per peak region across all the samples
count.matrix <- apply(log2.count.matrix, MARGIN = 1, FUN = function(X) (X - mean(X))/sd(X))
count.matrix <- t(count.matrix)

k <- 6
set.seed(081385)
km <- kmeans(count.matrix, centers = k, iter.max = 500) 
m.kmeans <- cbind(count.matrix, km$cluster)
o <- order(m.kmeans[,ncol(m.kmeans)])
m.kmeans <- m.kmeans[o,]
annotation_row <- data.frame(row.names = row.names(m.kmeans), km = as.factor(m.kmeans[,ncol(m.kmeans)])) 
ordered_mat <- m.kmeans[,1:ncol(m.kmeans)-1]

pheatmap::pheatmap(ordered_mat, 
                   scale = "none", 
                   show_rownames = FALSE, 
                   fontsize = 10, 
                   cellwidth = 10,  
                   main = "DAPs across all contrasts",  
                   annotation_col = annot_col, 
                   viridis::viridis(50), 
                   silent = FALSE , 
                   cluster_cols = FALSE, 
                   cluster_rows = FALSE, 
                   gaps_col = cumsum(c(0, table(decoder.data$time)))[-1],
                   annotation_row=annotation_row)

```

```{r save_peaks, eval=T}
df = ordered_mat
df= merge(df, annotation_row, by="row.names")
colnames(df)[1] = "peak"

spl <- strsplit(gsub("-|:", ".", df$peak), ".", fixed = TRUE)
df$chr = sapply(spl, "[", 1)
df$start =sapply(spl, "[", 2)
df$end = gsub("\\_.*","",sapply(spl, "[", 3))
row.names(df) = df$peak
df = df[row.names(ordered_mat),]

if (!all(row.names(annotation_row) == row.names(df))) {
  stop("Error: Row names in 'annotation_row' and 'df' are not identical.")
}

write.table(data.frame(df[,c("chr", "start", "end")], ID=df$peak, k=df$k), "ext_fig7b_atacseq_heatmap.bed", sep="\t", quote=F, col.names=F, row.names = F)
```

# BigWig Heatmap

Heatmap of the above.
Bins with read counts greater than the 75th percentile + 1.5*IQR were capped at that value.


```{r onCluster, eval=F}
# cd /home/paz2005/archive/2023_01_mary_philip/figures_for_paper_with2017
library(profileplyr)
library(data.table)
library(BiocParallel)
bigwigfiles =  c("N.bw", "E6hr.bw", "E12hr.bw", "E24hr.bw", "E5.bw", "E7.bw", "M.bw")
all_profiles =  BamBigwig_to_chipProfile(signalFiles = bigwigfiles,
                                         testRanges ="./ext_fig7b_atacseq_heatmap.bed",
                                         style = "point",
                                         format = "bigwig",
                                         distanceAround = 1000, 
                                         quant_params=MulticoreParam(14))

assays <- lapply(1:length(bigwigfiles), function(i) assay(all_profiles, i))

# check if row names identical 
identical_row_names <- all(sapply(assays, function(x) identical(row.names(x), row.names(assays[[1]]))))

if(identical_row_names) {
  merged <- do.call(cbind, assays)
  rownames(merged) <-  rowData(all_profiles)$name
} else {
  print("Error: row names are not identical for all assays")
  merged <- NULL
}

saveRDS(merged, "ext_fig7b_atacseq_heatmap.RDS")
```



```{r hm_clsrow_dend_reorder_f_row_annot, fig.width=5, fig.height=7, fig.path='ext_fig7b_atacseq_heatmap/', dev=c('png')}
all = readRDS("ext_fig7b_atacseq_heatmap.RDS")
peaks = df
all = all[df$peak,]

# To improve visibility, bins with read counts greater than the 75th percentile + 1.5*IQR were capped at that value.
all[all >= as.numeric((quantile(all, c(0, .75))+IQR(all)*1.5)[2])] <- as.numeric((quantile(all, c(0, .75))+IQR(all)*1.5)[2])
colnames(all)  = seq(1:ncol(all))

names =   c("N.bw", "E6hr.bw", "E12hr.bw", "E24hr.bw", "E5.bw", "E7.bw", "M.bw")
annot_col = data.frame(cond =rep(gsub(".bw", "", names), each=100),row.names=colnames(all))
annot_col$cond = factor(annot_col$cond, levels=unique(rep(gsub(".bw", "", names), each=100)))

table(df$peak == row.names(all))

ha <- ComplexHeatmap::HeatmapAnnotation(
  foo = ComplexHeatmap::anno_block(
    gp = gpar(fill = c("#636363", "#ccece6", "#99d8c9", "#66c2a4", "#238b45", "#238b45", "#00441b")),
    labels = unique(rep(gsub(".bw", "", names), each = 100)),
    labels_gp = gpar(col = rep("white", 7))
  )
)

hm = ComplexHeatmap::Heatmap(
  all,
  border = FALSE,
  cluster_rows = TRUE,
  cluster_columns = FALSE,
  col = rev(brewer.pal(n = 11, name = 'RdBu')),
  show_row_names = FALSE,
  heatmap_legend_param = list(direction = "horizontal", title = NULL, legend_width = unit(6, "cm")),
  column_split = annot_col$cond,
  top_annotation = ha,
  column_title = NULL,
  show_column_names = FALSE,
  use_raster = FALSE,
  show_row_dend = FALSE,
  row_split = df$km,
  cluster_row_slices = FALSE,
  row_title = NULL,
  row_gap = unit(0, "mm"),
  row_dend_reorder = FALSE
)
ComplexHeatmap::draw(hm, heatmap_legend_side = "bottom")
```

# Session Info
```{r session, message=FALSE, warning=FALSE, cache=FALSE,echo=FALSE, fig.width=10, fig.height=5.5, context="data"}
sessionInfo()
```