---
title: Ext. Fig6
subtitle:  "RNA expression of genes with enriched STAT1 binding motifs from effector or tumor conditions across timepoints. Heatmaps are Z-score normalized across rows. Dots represent genes of interest"
author: "pz"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    highlight: tango
    df_print: paged
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
    code_folding: hide
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, result='hold',fig.width=10,tidy=TRUE, cache=FALSE)
knitr::opts_knit$set(progress=TRUE,verbose=TRUE)
```

```{r libraries, cache=FALSE}
library(data.table)
library(magrittr)
library(knitr)
library(ggplot2); theme_set(theme_bw(base_size = 16))
library(DESeq2)
library(tidyverse)
library(chromVAR)
library(chromVARmotifs)
library(motifmatchr)
library(BSgenome.Mmusculus.UCSC.mm10)
library(pheatmap)
library(viridis)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(RColorBrewer)
```



```{r decoder}
decoderFile <- "atac_decoder.csv"
decoder.data <- read.table(decoderFile,header=T,stringsAsFactors=F, sep=",")
decoder.data$SampleID = make.names(decoder.data$SampleID)
decoder.data$Condition = factor(decoder.data$Condition)
# select early timepoints only
decoder.data = subset(decoder.data, (Condition %in% c("E6hr", "E12hr", "E24hr", "L6hr", "L12hr", "L1d")))
```

# Samples

The following samples were part of this analysis:

```{r samples}
as_tibble(decoder.data)
```

```{r loadDESeq2Object_extractRawCounts}
# see runDiffBind.R, runSVA.R
load("sva_dds_vsd.RData")

# load counts
df = assay(dds)[,decoder.data$SampleID] %>% as.data.frame()

spl <- strsplit(gsub("-|:", ".", row.names(df)), ".", fixed = TRUE)
df$seqnames = sapply(spl, "[", 1)
df$start =sapply(spl, "[", 2)
df$end = gsub("\\_.*","",sapply(spl, "[", 3))
  
gr = GRanges(df)

if (!all(gr$peak == row.names(df))) {
  stop("Row names do not match GRanges object")
}

if (!all(colnames(mcols(gr)) == decoder.data$SampleID)) {
  stop("Column names do not match decoder data SampleID")
}

se = SummarizedExperiment(assays = list(counts = Matrix::Matrix(as.matrix(mcols(gr)))), 
                          colData=decoder.data, 
                          rowRanges=gr)
```

```{r readInRNAseqCounts}
# we want to filter out motif atlas for TFs that are expressed 
RNA.counts <- read.table(file = "../rnaseq/gene.counts.txt", header = TRUE, check.names=FALSE, row.names=1)
RNA.decoder.data <- fread("../rnaseq/decoder.txt") %>% as.data.frame()
RNA.decoder.data = subset(RNA.decoder.data, group != "N") # drop naive 
RNA.counts <- RNA.counts[,RNA.decoder.data$sample.ID]
# compute cpm 
cpms = edgeR::cpm(RNA.counts)
# remove genes that don't have at least 5 CPM on average in each rowMean
cpms_filt = cpms[rowMeans(cpms) > 5,]
```

```{r}
motifs_expr <- mouse_pwms_v2
motif_names = ArchR:::.summarizeChromVARMotifs(motifs_expr)$motifSummary %>% as.data.frame()
# only retain motifs that are expressed in the RNA-seq data
motif_names = motif_names[motif_names$name %in% row.names(cpms_filt),]
motifs_expr = motifs_expr[which(names(motifs_expr) %in% motif_names$ID)]
motifs <- motifs_expr
```

# chromVAR

```{r}
# Filter rows with total counts > 5
counts <- se[rowSums(assay(se)) > 5, ]

# Compute deviations and variability
set.seed(081385)
counts <- chromVAR::addGCBias(counts, genome = BSgenome.Mmusculus.UCSC.mm10)
set.seed(081385)
motif_ix <- motifmatchr::matchMotifs(motifs, rowRanges(counts), genome = BSgenome.Mmusculus.UCSC.mm10)
set.seed(081385)
dev <- chromVAR::computeDeviations(object = counts, annotations = motif_ix)
set.seed(081385)
variability <- chromVAR::computeVariability(dev)

res_df <- cbind(variability, rank = rank(-1 * variability$variability, ties.method = "random"))
res_df <- res_df[order(res_df$rank, decreasing = F), ]
res_df$rank <- NULL
res_df = res_df[res_df$p_value_adj < 0.05,]

# Filter top significant motifs
res_df <- cbind(variability, rank = rank(-1 * variability$variability, ties.method = "random"))
res_df <- res_df[order(res_df$rank, decreasing = F), ]
res_df <- res_df[res_df$p_value_adj < 0.05,]
n <- 30
topVariable <- res_df[1:n, ]

# Replace "LINE" in the name with a suffix of the original row name
if (any(grepl("LINE", topVariable$name))) {
  topVariable[grep("LINE", topVariable$name), "name"] <- paste0("LINE-", sapply(strsplit(gsub("_", "-", row.names(topVariable[grep("LINE", topVariable$name), ])), "-"), `[`, 4))
}

# Merge top variable motifs with deviation scores
devTop <- merge(topVariable[, 1, drop = FALSE], assay(dev, "z"), by = 0)
devToPlot <- as.matrix(devTop[, -c(1:2)])
rownames(devToPlot) <- devTop[, 2]
```

## Top 30 most variable motifs

```{r chromVar_top30,  fig.width=8}
as_tibble(res_df)

col_order = c("E6-3","E6-2","E6-1","E12-3","E12-2","E12-1","E24-3","E24-2","E24-1","L6-3","L6-2","L6-1","L12-3", "L12-2", "L12-1", "L1d-3","L1d-2","L1d-1")

devToPlot_ordered = devToPlot[,make.names(col_order)]

annot_col = data.frame( colData(dev))[,"Condition", drop=F]

annotation_colors = list(Condition=c("E6hr"="#ccece6",
                                     "E12hr"="#99d8c9",
                                     "E24hr" =  "#66c2a4",
                                     "L6hr" = "#c6dbef",
                                     "L12hr" = "#6baed6",
                                     "L1d" = "#439dcb"))
pheatmap(devToPlot_ordered,
         scale = "none",
         annotation_col = annot_col,
         annotation_colors = annotation_colors,
         cluster_cols = FALSE,
         cluster_rows = TRUE,
         show_rownames = TRUE,
         clustering_method = "ward.D2",
         color = viridis(50),
         fontsize_row = 12,
         gaps_col = 9,
         silent = FALSE,
         cellwidth = 15,
         cellheight = 10,
         treeheight_col = 45,
         main = "",
         border_color = NA,
         treeheight_row = 0)
```

# Differential accessibility analysis

The following peaks were detected as differentially accessible:


```{r da, eval=T,echo=FALSE, message=FALSE, warning=FALSE,   context="data"}
varInt="Condition"
dbs.all <- list()
alphaTR = 0.05
lfcThres = 1


comps = c("E6hr_vs_L6hr","E12hr_vs_L12hr","E24hr_vs_L1d")
for (comp in comps){
  levelRef <- strsplit(comp, "_vs_")[[1]][1]
  levelTest <- strsplit(comp, "_vs_")[[1]][2]
  dbs.all[[paste0(levelTest,"_vs_",levelRef)]] <- 
    results(dds, contrast=c(varInt, levelTest, levelRef), alpha=alphaTR) %>% as.data.frame()
}
names(dbs.all) = gsub("hr", "", names(dbs.all))
```

The following genes were detected as differentially expressed (adjusted p < `r alphaTR` and abs(log2FoldChange) > `r lfcThres`):

```{r printDARtable}
dbs.all.sig <- lapply(names(dbs.all), function(x){
  subset(dbs.all[[x]], padj < alphaTR & abs(log2FoldChange) > lfcThres)
   })
names(dbs.all.sig) <- names(dbs.all)

res = data.frame(sapply(dbs.all.sig, NROW))
colnames(res) <-  paste0("padj<",alphaTR, " & ","abs(log2FoldChange)>",lfcThres)

rel.contrs = 1:nrow(res)

as_tibble(cbind(res,"contrast #"=rel.contrs) %>%  tibble::rownames_to_column(var="contrast"))
```



# Differential expression analysis


```{r loadRnaSeq}
decoderFile <- "../rnaseq/decoder.txt"
decoder.data <- fread(decoderFile) %>% as.data.frame()
decoder.data$group <- factor(decoder.data$group, levels = c("N", "E6", "E12", "E24", "L6", "L12", "L24"))
decoder.data$type <- factor(decoder.data$type, levels = c("naive", "effector", "tumor"))
decoder.data$time <- factor(decoder.data$time, levels = c("0hr", "6hr", "12hr", "24hr"))
decoder.data$replicate <- factor(decoder.data$replicate)
decoder.data = decoder.data[order(decoder.data$group),]
row.names(decoder.data) = decoder.data$sample.ID
counts <- read.table(file = "../rnaseq/gene.counts.txt", header = TRUE, check.names=FALSE, row.names=1)
decoder.data <- decoder.data[decoder.data$sample.ID %in% colnames(counts),]

if (!identical(colnames(counts), decoder.data$sample.ID)) {
  stop("Column names in counts do not match sample IDs in decoder data")
}

counts <- counts[,c(decoder.data$sample.ID)]
dds <- DESeq2::DESeqDataSetFromMatrix(countData = counts, 
                                      colData = decoder.data, 
                                      design = ~ group)

keep <- edgeR::filterByExpr(counts(dds), design = model.matrix(~ group, data = colData(dds)))
dds <- dds[keep,]

dds <- estimateSizeFactors(dds)
vsd <- DESeq2::vst(dds, blind = TRUE)
normCounts = assay(vsd)
dds <- DESeq(dds)
```

```{r de, eval=T,echo=FALSE, message=FALSE, warning=FALSE,   context="data"}
varInt="group"
de.all <- list()
alphaTR = 0.05
lfcThres = 1


comps = c("L6_vs_E6","L12_vs_E12","L24_vs_E24")
for (comp in comps){
  levelRef <- strsplit(comp, "_vs_")[[1]][2]
  levelTest <- strsplit(comp, "_vs_")[[1]][1]
  de.all[[paste0(levelTest,"_vs_",levelRef)]] <- 
    results(dds, contrast=c(varInt, levelTest, levelRef), alpha=alphaTR) %>% as.data.frame()
}
```

The following genes were detected as differentially expressed (adjusted p < `r alphaTR` and abs(log2FoldChange) > `r lfcThres`):


```{r printDEGtable}
de.all.sig <- lapply(de.all, subset, padj < alphaTR & abs(log2FoldChange) > lfcThres)

res =  as.data.frame(sapply(de.all.sig, nrow))
colnames(res) <- paste0("padj<",alphaTR, " & ","abs(log2FoldChange)>",lfcThres)

rel.contrs = 1:nrow(res)

as_tibble(cbind(res,"contrast #"=rel.contrs) %>%  tibble::rownames_to_column(var="contrast"))
```

# Heatmap of genes with enriched STAT1 {.tabset}

What are we doing?

- Get the index of the motif
- Get the indices of the genomic regions that correspond to the motif
- Subset those regions to only include those that were differentially accessible (up_in_L or up_in_E)
- Annotate these genomic regions using the ChIPseeker w.r.t Gencode vM25 -- exclude regions that map to distal intergenic
- Now that we have a list of peaks that i) are associated with a TF, ii) are differentially accessible (up_in_L or up_in_E), and iii) annotated with repspect to a gene, subset to those genes that are also differentially expressed (in any direction)
- Make heatmaps of those genes associated with the TF stratified by ATAC-seq direction ( (up_in_L or up_in_E))


```{r}
atac_up_in_L = lapply(names(dbs.all.sig), function(x){
  subset(dbs.all.sig[[x]], padj < alphaTR & log2FoldChange > 0) %>% row.names(.)
   }) %>% unlist() %>% unique()

atac_up_in_E = lapply(names(dbs.all.sig), function(x){
  subset(dbs.all.sig[[x]], padj < alphaTR &  log2FoldChange < 0 ) %>% row.names(.)
   }) %>% unlist() %>% unique()

rna_up_in_L = lapply(names(de.all.sig), function(x){
  subset(de.all.sig[[x]], padj < alphaTR & log2FoldChange > 0) %>% row.names(.)
   }) %>% unlist() %>% unique()

rna_up_in_E = lapply(names(de.all.sig), function(x){
  subset(de.all.sig[[x]], padj < alphaTR &  log2FoldChange < 0 ) %>% row.names(.)
   }) %>% unlist() %>% unique()
```

```{r fig.width=7, fig.height=10,results="asis"}
#genecode.txdb<- makeTxDbFromGFF("gencode.vM25.annotation.gtf", format="gtf", organism = "Mus musculus")
#saveDb(genecode.txdb, 'txdb.gencodevM25.sqlite')
genecode.txdb = loadDb("txdb.gencodevM25.sqlite")
genecode.dt =  rtracklayer::import("gencode.vM25.annotation.gene.gtf")
genecode.df = as.data.frame(genecode.dt)
genecode.df = genecode.df[,c("gene_id", "gene_name")]
colnames(genecode.df) = c("geneId", "symbol")

# Subset devTop by TFs of interest
devTop_subset <- devTop[grep("Stat1", devTop$name, value=FALSE),]

# Iterate over the subset and print a header for each iteration
lapply(devTop_subset$Row.names, function(y) {
#  cat("\n\n##", y, "\n\n")

  # Get the index of the motif
  idx_of_motif <- which(colnames(motif_ix) == y)

  # Get the indices of the rows corresponding to the motif
  idx <- row.names(assay(motif_ix)[assay(motif_ix)[, idx_of_motif], ])

  # Get the ranges corresponding to the rows
  counts <- se[rowSums(assay(se)) > 5, ]
  ranges <- rowRanges(counts)[idx, ]
  ranges$peak <- paste0(as.data.frame(ranges)$seqnames, ":", as.data.frame(ranges)$start, "-", as.data.frame(ranges)$end)
  ranges$motif <- y

  # Annotate the peaks with ChIPseeker
  ranges_annot <- ChIPseeker::annotatePeak(ranges, tssRegion=c(-1000, 1000), TxDb=genecode.txdb, annoDb="org.Mm.eg.db", verbose=FALSE)
  ranges_annot_df <- as.data.frame(ranges_annot)

  # Subset the annotation dataframe
  ranges_annot_df <- subset(ranges_annot_df, annotation != "Distal Intergenic")
  ranges_annot_df <- merge(ranges_annot_df, genecode.df, by="geneId")

  # Subset ranges_annot_df by matching peaks with atac_up_in_L and atac_up_in_E
  up_in_L_atac_merged <- ranges_annot_df[which(ranges_annot_df$peak %in% atac_up_in_L), ]
  up_in_E_atac_merged <- ranges_annot_df[which(ranges_annot_df$peak %in% atac_up_in_E), ]

  # Set the order of columns and create a dataframe for annotations
  col_order <- c("E6", "E12", "E24", "L6", "L12", "L24")
  decoder.data.sub = subset(decoder.data, type != "naive")
  annot_col <- data.frame(cond=decoder.data.sub$group, row.names=decoder.data.sub$sample.ID)
  annot_col$cond <- ordered(annot_col$cond, levels=col_order)
  annot_col <- annot_col[order(annot_col$cond), , drop=FALSE]

    cat("\n\n##", "Up in tumor", "\n\n")
  # Get the log2 count matrix for the selected genes
  log2.count.matrix.sub <- normCounts[unique(up_in_L_atac_merged$symbol)[unique(up_in_L_atac_merged$symbol) %in% unique(c(rna_up_in_L, rna_up_in_E))], decoder.data.sub$sample.ID]
  
  # Define color palette
  my_colour <- list(cond=c("E6"="#ccece6", "E12"="#99d8c9", "E24"="#66c2a4", "L6"="#c6dbef", "L12"="#6baed6", "L24"="#439dcb"))

  # Create the heatmaps
  kclustcols <- colorRampPalette(rev(brewer.pal(n=5, name="RdBu")))(100)
  p1 <- ComplexHeatmap::pheatmap(log2.count.matrix.sub[, row.names(annot_col)], 
                                 scale = "row", 
                                 show_rownames = TRUE, 
                                 fontsize = 10, 
                                 main = paste0(y, "\nup in ATAC L"),  
                                 silent = FALSE , 
                                 cluster_cols = FALSE, 
                                 cluster_rows = TRUE, 
                                 annotation_col = annot_col, 
                                 color = kclustcols, 
                                 border_color = NA, 
                                 cellwidth = 20, 
                                 cellheight = 7, 
                                 treeheight_row = 0, 
                                 show_row_dend = FALSE, 
                                 gaps_col = c(9), 
                                 annotation_colors = my_colour, 
                                 fontsize_row = 8,
                                 show_colnames=FALSE)
  print(p1)
  
  cat("\n\n##", "Up in infection", "\n\n")
  log2.count.matrix.sub = normCounts[unique(up_in_E_atac_merged$symbol)[unique(up_in_E_atac_merged$symbol) %in% unique(c(rna_up_in_L, rna_up_in_E))],decoder.data.sub$sample.ID]
  
  p2 <- ComplexHeatmap::pheatmap(log2.count.matrix.sub[, row.names(annot_col)], 
                                 scale = "row", 
                                 show_rownames = TRUE, 
                                 fontsize = 10, 
                                 main = paste0(y, "\nup in ATAC E"),  
                                 silent = FALSE , 
                                 cluster_cols = FALSE, 
                                 cluster_rows = TRUE, 
                                 annotation_col = annot_col, 
                                 color = kclustcols, 
                                 border_color = NA, 
                                 cellwidth = 20, 
                                 cellheight = 7, 
                                 treeheight_row = 0, 
                                 show_row_dend = FALSE, 
                                 gaps_col = c(9), 
                                 annotation_colors = my_colour, 
                                 fontsize_row = 8,
                                 show_colnames=FALSE)
  print(p2)
}) %>% invisible()
```

# Session Info
```{r session, message=FALSE, warning=FALSE, cache=FALSE,echo=FALSE, fig.width=10, fig.height=5.5, context="data"}
sessionInfo()
```