---
title: Fig5
subtitle:  "Chromatin accessibility heatmap."
author: "pz"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    highlight: tango
    df_print: paged
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
    code_folding: hide
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, result='hold',fig.width=10,tidy=TRUE, cache=FALSE)
knitr::opts_knit$set(progress=TRUE,verbose=TRUE)
```

```{r libraries, cache=FALSE}
library(data.table)
library(magrittr)
library(knitr)
library(ggplot2); theme_set(theme_bw(base_size = 16))
library(DESeq2)
library(tidyverse)
library(eulerr)
library(profileplyr)
library(grid)
library(ComplexHeatmap)
library(RColorBrewer)
```



```{r decoder}
decoderFile <- "atac_decoder.csv"
decoder.data <- read.table(decoderFile,header=T,stringsAsFactors=F, sep=",")
decoder.data$SampleID = make.names(decoder.data$SampleID)
decoder.data$Condition = factor(decoder.data$Condition)
decoder.data$Condition <- relevel(decoder.data$Condition, ref = "N")
```


```{r loadDESeq2Object}
# see runDiffBind.R, runSVA.R
load("sva_dds_vsd.RData")
normCounts = assay(vsd)
```

# Samples

The following samples were part of this analysis:

```{r samples}
as_tibble(decoder.data)
```

# Differential accessibility analysis


```{r da}
varInt="Condition"
dbs.all <- list()
alphaTR = 0.05
lfcThres = 1.5

comps = c("L6hr_vs_N", "L12hr_vs_N", "L1d_vs_N", "E6hr_vs_N","E12hr_vs_N", "E24hr_vs_N")
for (comp in comps){
  levelTest <- strsplit(comp, "_vs_")[[1]][1]
  levelRef <- strsplit(comp, "_vs_")[[1]][2]
  dbs.all[[paste0(levelTest,"_vs_",levelRef)]] <- 
    results(dds, contrast=c(varInt, levelTest, levelRef), alpha=alphaTR) %>% as.data.frame()
}

names(dbs.all) = gsub("hr", "", names(dbs.all))
```


The following genes were detected as differentially expressed (adjusted p < `r alphaTR` and abs(log2FoldChange) > `r lfcThres`):

```{r printDEGtable}
dbs.all.sig <- lapply(names(dbs.all), function(x){
  subset(dbs.all[[x]], padj < alphaTR & abs(log2FoldChange) > lfcThres)
   })
names(dbs.all.sig) <- names(dbs.all)

res = data.frame(sapply(dbs.all.sig, NROW))
colnames(res) <-  paste0("padj<",alphaTR, " & ","abs(log2FoldChange)>",lfcThres)

rel.contrs = 1:nrow(res)

as_tibble(cbind(res,"contrast #"=rel.contrs) %>%  tibble::rownames_to_column(var="contrast"))
```

# Venn diagram

venn betwen union(E6 vs. N; E12 vs. N; E24 vs. N) vs.  union(L6 vs. N; L12 vs. N; L24 vs. N)
 

```{r euler}
ll = list("N->L" = Reduce(union, list(row.names(dbs.all.sig$L6_vs_N),row.names( dbs.all.sig$L12_vs_N), row.names(dbs.all.sig$L1d_vs_N))),
          "N->E" =  Reduce(union, list(row.names(dbs.all.sig$E6_vs_N),row.names( dbs.all.sig$E12_vs_N), row.names(dbs.all.sig$E24_vs_N))))

p = plot(euler(ll),
     fills = c("#c6dbef", "#ccece6"),
     edges = T,
     fontsize = 10,
     quantities = list(fontsize = 12), cex.axis=1)
print(p)
```


```{r  save_peaks_in_venn}
ll = list("N->L" = Reduce(union, list(row.names(dbs.all.sig$L6_vs_N),row.names( dbs.all.sig$L12_vs_N), row.names(dbs.all.sig$L1d_vs_N))),
          "N->E" =  Reduce(union, list(row.names(dbs.all.sig$E6_vs_N),row.names( dbs.all.sig$E12_vs_N), row.names(dbs.all.sig$E24_vs_N))))
l_union = Reduce(union, list(row.names(dbs.all.sig$L6_vs_N),row.names( dbs.all.sig$L12_vs_N), row.names(dbs.all.sig$L1d_vs_N)))
e_union = Reduce(union, list(row.names(dbs.all.sig$E6_vs_N),row.names( dbs.all.sig$E12_vs_N), row.names(dbs.all.sig$E24_vs_N)))

inter =  Reduce(intersect, ll)
in_l6 = setdiff(l_union, e_union)
in_e6 = setdiff(e_union,l_union)

all.sig.rows = c(in_e6, inter, in_l6)

df =  normCounts[all.sig.rows, decoder.data$SampleID]

df = cbind(df, data.frame(cls=c(rep("in_e6",length(in_e6)), rep("inter",length(inter)), rep("in_l6",length(in_l6)))))
df$peak = row.names(df)

spl <- strsplit(gsub("-|:", ".", df$peak), ".", fixed = TRUE)
df$chr = sapply(spl, "[", 1)
df$start =sapply(spl, "[", 2)
df$end = gsub("\\_.*","",sapply(spl, "[", 3))
df$k = ifelse(df$cls == "in_e6", 1, ifelse(df$cls == "inter", 2,3))
df = df[order(df$k),]
#table(df$cls,df$k)

write.table(data.frame(df[,c("chr", "start", "end")], ID=df$peak, k=df$k), "fig5_atacseq_heatmap.bed", sep="\t", quote=F, col.names=F, row.names = F)
```

# Heatmap

Heatmap of the above venn diagram.
Bins with read counts greater than the 75th percentile + 1.5*IQR were capped at that value.


```{r onCluster, eval=F}
library(profileplyr)
library(data.table)
library(BiocParallel)
peaks = fread("fig5_atacseq_heatmap.bed")
bigwigfiles =  c("N.bw", "E6.bw", "E12.bw", "E24.bw", "L6.bw", "L12.bw", "L24.bw")
all_profiles =  BamBigwig_to_chipProfile(signalFiles = bigwigfiles,
                                         testRanges ="./fig5_atacseq_heatmap.bed",
                                         style = "point",
                                         format = "bigwig",
                                         distanceAround = 1000, 
                                         quant_params=MulticoreParam(12))

assays <- lapply(1:length(bigwigfiles), function(i) assay(all_profiles, i))

# check if row names identical 
identical_row_names <- all(sapply(assays, function(x) identical(row.names(x), row.names(assays[[1]]))))

if(identical_row_names) {
  merged <- do.call(cbind, assays)
  rownames(merged) <-  rowData(all_profiles)$name
} else {
  print("Error: row names are not identical for all assays")
  merged <- NULL
}

saveRDS(merged, "fig5_atacseq_heatmap.RDS")
```



```{r hm_clsrow_dend_reorder_f_row_annot,  fig.width=4.5, fig.height=8}
all = readRDS("fig5_atacseq_heatmap.RDS")
peaks = df
all = all[df$peak,]
# To improve visibility, bins with read counts greater than the 75th percentile + 1.5*IQR were capped at that value.
all[all >= as.numeric((quantile(all, c(0, .75))+IQR(all)*1.5)[2])] <- as.numeric((quantile(all, c(0, .75))+IQR(all)*1.5)[2])
colnames(all)  = seq(1:ncol(all))
names =  c("N.bw", "E6.bw", "E12.bw", "E24.bw", "L6.bw", "L12.bw", "L24.bw")
annot_col = data.frame(cond =rep(gsub(".bw", "", names), each=100),row.names=colnames(all))
annot_col$cond = factor(annot_col$cond, levels=unique(rep(gsub(".bw", "", names), each=100)))

table(df$peak == row.names(all))

df$cls = factor(df$cls, levels= c("in_e6", "inter", "in_l6"))

ha <- ComplexHeatmap::HeatmapAnnotation(
  foo = ComplexHeatmap::anno_block(
    gp = gpar(fill = c("#636363", "#ccece6","#99d8c9", "#66c2a4", "#c6dbef", "#6baed6", "#4398cb")),
    labels = unique(rep(gsub(".bw", "", names), each = 100)),
    labels_gp = gpar(col = c("white", "white", "white", "white", "white", "white", "white"))
  )
)

annoRow=list("cls" = c(structure(c("#218a42","#7030A0","#2171B5"), names =  c("in_e6", "inter", "in_l6"))))
cls_levels = c("in_e6", "inter", "in_l6")

hm <- ComplexHeatmap::Heatmap(all,
  border = F,
  cluster_rows = T,
  cluster_columns = F,
  col = rev(brewer.pal(n = 11, name ='RdBu')),
  show_row_names = F,
  heatmap_legend_param = list(
    direction = "horizontal",
    title = NULL,
    legend_width = unit(6, "cm")
  ),
  column_split = annot_col$cond,
  top_annotation = ha,
  column_title = NULL,
  show_column_names = F,
  use_raster = F,
  show_row_dend = F,
  row_split = df$cls,
  cluster_row_slices = F,
  row_title = NULL,
  row_gap = unit(0.5, "mm"),
  row_dend_reorder = F,
  left_annotation = rowAnnotation(
    annotation_legend_param = list(
      cls = list(labels = cls_levels, at = cls_levels)
    ),
    df = df[, c("cls"), drop = F],
    show_legend = FALSE,
    col = annoRow,
    show_annotation_name = FALSE,
    simple_anno_size = unit(7, "bigpts")
  )
)

ComplexHeatmap::draw(hm, heatmap_legend_side="bottom")
```

# Session Info
```{r session, message=FALSE, warning=FALSE, cache=FALSE,echo=FALSE, fig.width=10, fig.height=5.5, context="data"}
sessionInfo()
```