---
title: Fig 7
subtitle:  "Chromatin accessibility heatmap."
author: "pz"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    highlight: tango
    df_print: paged
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
    code_folding: hide
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, result='hold',fig.width=10,tidy=TRUE, cache=FALSE)
knitr::opts_knit$set(progress=TRUE,verbose=TRUE)
```

```{r libraries, cache=FALSE}
library(data.table)
library(magrittr)
library(knitr)
library(ggplot2); theme_set(theme_bw(base_size = 16))
library(DESeq2)
library(tidyverse)
library(eulerr)
library(profileplyr)
library(grid)
library(ComplexHeatmap)
library(RColorBrewer)
```



```{r decoder}
decoderFile <- "atac_decoder.csv"
decoder.data <- read.table(decoderFile,header=T,stringsAsFactors=F, sep=",")
decoder.data$SampleID = make.names(decoder.data$SampleID)
decoder.data$Condition = factor(decoder.data$Condition)
decoder.data$Condition <- relevel(decoder.data$Condition, ref = "N")
decoder.data = subset(decoder.data, (Condition %in% c("N", "L10d" ,"P10d","L1d", "P1d", "L5d", "P5d", "M")))
decoder.data = subset(decoder.data, batch == "m2023" | Condition == "M")
decoder.data$Condition <- relevel(decoder.data$Condition, ref = "N")
decoder.data$Condition <- factor(decoder.data$Condition, levels = c("N", "M", "L1d", "L5d", "L10d", "P1d", "P5d", "P10d"))
decoder.data = with(decoder.data, decoder.data[order(Condition),])
```


```{r loadDESeq2Object}
# see runDiffBind.R, createParkingDDSObject_noSVA.R
load("parking_dds.RData")
vsd <- DESeq2::vst(dds, blind=TRUE)
normCounts = assay(vsd)
baseMeanPerLvl <- sapply( levels(vsd$Condition), function(lvl) rowMeans( assay(vsd)[,vsd$Condition == lvl, drop=F]))
```

# Samples

The following samples were part of this analysis:

```{r samples}
as_tibble(decoder.data)
```

# Differential accessibility analysis


```{r da}
varInt="Condition"
dbs.all <- list()
alphaTR = 0.05
lfcThres = 1.5

comps = c("N_vs_P1d", "N_vs_P5d", "N_vs_P10d",
          "N_vs_L1d", "N_vs_L5d", "N_vs_L10d",
          "M_vs_P1d", "M_vs_P5d", "M_vs_P10d",
          "M_vs_L1d", "M_vs_L5d", "M_vs_L10d",
          "L1d_vs_P1d", "L5d_vs_P5d", "L10d_vs_P10d",
          "P1d_vs_P5d", "P5d_vs_P10d", "P1d_vs_M",
          "P5d_vs_M", "P10d_vs_M")
for (comp in comps){
  levelRef <- strsplit(comp, "_vs_")[[1]][1]
  levelTest <- strsplit(comp, "_vs_")[[1]][2]
  dbs.all[[paste0(levelTest,"_vs_",levelRef)]] <- 
    results(dds, contrast=c(varInt, levelTest, levelRef), alpha=alphaTR) %>% as.data.frame()
}
names(dbs.all) = gsub("hr", "", names(dbs.all))
```


The following genes were detected as differentially expressed (adjusted p < `r alphaTR` and abs(log2FoldChange) > `r lfcThres`):

```{r printDEGtable}
dbs.all.sig <- lapply(names(dbs.all), function(x){
  subset(dbs.all[[x]], padj < alphaTR & abs(log2FoldChange) > lfcThres)
   })
names(dbs.all.sig) <- names(dbs.all)

res = data.frame(sapply(dbs.all.sig, NROW))
colnames(res) <-  paste0("padj<",alphaTR, " & ","abs(log2FoldChange)>",lfcThres)

rel.contrs = 1:nrow(res)

as_tibble(cbind(res,"contrast #"=rel.contrs) %>%  tibble::rownames_to_column(var="contrast"))
```

# Heatmap infection (k = 6)


```{r}
all.sig.rows <- unique(unlist(lapply(dbs.all.sig, row.names)))

annot_col <- data.frame(cond = decoder.data$Condition, row.names = make.names(decoder.data$SampleID))
log2.count.matrix <- normCounts[all.sig.rows, make.names(decoder.data$SampleID)]

#Obtain the z-score per peak region across all the samples
count.matrix <- apply(log2.count.matrix, MARGIN = 1, FUN = function(X) (X - mean(X))/sd(X))
count.matrix <- t(count.matrix)

k <- 6
set.seed(081385)
km <- kmeans(count.matrix, centers = k, iter.max = 500) 
m.kmeans <- cbind(count.matrix, km$cluster)
o <- order(m.kmeans[,ncol(m.kmeans)])
m.kmeans <- m.kmeans[o,]
annotation_row <- data.frame(row.names = row.names(m.kmeans), km = as.factor(m.kmeans[,ncol(m.kmeans)])) 
ordered_mat <- m.kmeans[,1:ncol(m.kmeans)-1]

pheatmap::pheatmap(ordered_mat, 
                   scale = "none", 
                   show_rownames = FALSE, 
                   fontsize = 10, 
                   cellwidth = 10,  
                   main = "DAPs across all contrasts",  
                   annotation_col = annot_col, 
                   viridis::viridis(50), 
                   silent = FALSE , 
                   cluster_cols = FALSE, 
                   cluster_rows = FALSE, 
                   gaps_col = cumsum(c(0, table(decoder.data$Condition)))[-1],
                   annotation_row=annotation_row)

```

```{r save_peaks, eval=T}
df = ordered_mat
df= merge(df, annotation_row, by="row.names")
colnames(df)[1] = "peak"

spl <- strsplit(gsub("-|:", ".", df$peak), ".", fixed = TRUE)
df$chr = sapply(spl, "[", 1)
df$start =sapply(spl, "[", 2)
df$end = gsub("\\_.*","",sapply(spl, "[", 3))
row.names(df) = df$peak
df = df[row.names(ordered_mat),]

if (!all(row.names(annotation_row) == row.names(df))) {
  stop("Error: Row names in 'annotation_row' and 'df' are not identical.")
}

write.table(data.frame(df[,c("chr", "start", "end")], ID=df$peak, k=df$k), "fig7_atacseq_heatmap_parking.bed", sep="\t", quote=F, col.names=F, row.names = F)
```

# BigWig Heatmap

Heatmap of the above.
Bins with read counts greater than the 75th percentile + 1.5*IQR were capped at that value.


```{r onCluster, eval=F}
# cd /home/paz2005/archive/2023_01_mary_philip/figures_for_paper_parking
library(profileplyr)
library(data.table)
library(BiocParallel)
bigwigfiles =  c("N.bw", "M.bw", "L1d.bw", "L5d.bw", "L10d.bw", "P1d.bw", "P5d.bw","P10d.bw")

# the below code is stripped from the profileplyr::BamBigwig_to_chipProfile function -- the regualar function (below) does not work for some odd reason because of the P10d.bw file
# all_profiles =  BamBigwig_to_chipProfile(signalFiles = bigwigfiles, 
#                                          testRanges ="./fig7_atacseq_heatmap_parking.bed",
#                                          style = "point",
#                                          format = "bigwig",
#                                          distanceAround = 1000, 
#                                          quant_params=MulticoreParam(12))

signalFiles  = bigwigfiles
testRanges ="fig7_atacseq_heatmap_parking.bed"
style = "point"
format = "bigwig"
distanceAround = 1000
quant_params = MulticoreParam(9)
nOfWindows = 100
bin_size = 20
distanceUp = 1000
distanceDown = 1000
library(rtracklayer)

if (format == "bigwig") {
  add_chr <- function(genomeCov) {
    if (!(any(grepl("chr", names(genomeCov))))) {
      oddChrom <- grepl("GL", names(genomeCov))
      for (i in seq_along(names(genomeCov))) {
        if (!oddChrom[i]) {
          names(genomeCov)[i] <- paste0("chr", names(genomeCov)[i])
        }
        else {
          names(genomeCov)[i] <- names(genomeCov)[i]
        }
      }
    }
    return(genomeCov)
  }
  message("Loading bigwig files.")
  genomeCov_list <- lapply(signalFiles, import.bw, as = "RleList")
  genomeCov_list <- lapply(genomeCov_list, add_chr)
  genomeCov_names <- lapply(genomeCov_list, names)
  common_names <- Reduce(intersect, genomeCov_names)
  signalFiles_list <- lapply(genomeCov_list, function(x) x[names(x) %in% 
                                                             common_names])
  group_labels <- vector()
  testRanges_GR <- GRangesList()
  for (i in seq_along(testRanges)) {
    temp <- import(testRanges[i])
    seqlevelsStyle(temp) <- "UCSC"
    temp <- temp[seqnames(temp) %in% common_names]
    temp <- temp
    seqlevels(temp) <- seqlevelsInUse(temp)
    testRanges_GR[[i]] <- temp
    testRanges_GR[[i]]$sgGroup <- rep(basename(testRanges[i]), 
                                      length(testRanges_GR[[i]]))
    group_labels[i] <- basename(testRanges[i])
  }
  format <- "rlelist"
}


testRanges_GR_unlist <- unlist(testRanges_GR)
names(testRanges_GR_unlist) <- NULL
message("Making ChIPprofile object from signal files.")
BPPARAM <- quant_params
ChIPprofile_combined <- bplapply(signalFiles_list, 
                                 soGGi::regionPlot, testRanges = testRanges_GR_unlist, format = format, 
                                 style = style, nOfWindows = nOfWindows, distanceAround = distanceAround, 
                                 distanceUp = distanceUp, distanceDown = distanceDown,  BPPARAM = BPPARAM)

# Compare the row names of the first and eighth elements of ChIPprofile_combined
table(rownames(ChIPprofile_combined[[1]]) ==  rownames(ChIPprofile_combined[[8]]))

# Compare the row names of the first element of ChIPprofile_combined
# with the subset of the eighth element that corresponds to the row names of the first element
table( rownames(ChIPprofile_combined[[1]]) ==  rownames(ChIPprofile_combined[[8]][ rownames(ChIPprofile_combined[[1]]),]) )

# Update the eighth element of ChIPprofile_combined with a subset that corresponds to the row names of the first element
# the 8th file is P10d.bw -- he normal BamBigwig_to_chipProfile() function fails to merge  correctly becaue it's not in the same order 
ChIPprofile_combined[[8]] = ChIPprofile_combined[[8]][ rownames(ChIPprofile_combined[[1]]),]

ChIPprofile_for_proplyr <- do.call(c, ChIPprofile_combined)

if (is.character(signalFiles)) {
  metadata(ChIPprofile_for_proplyr)$names <- signalFiles
}
rowRanges(ChIPprofile_for_proplyr)$sgGroup <- factor(rowRanges(ChIPprofile_for_proplyr)$sgGroup,  levels = group_labels)
if (style == "point") {
  if (bin_size > 1) {
    bin_matrix <- function(matrix, bin_size) {
      bin_number <- as.integer(ncol(matrix)/bin_size)
      bin_mean_result <- list()
      for (i in seq(bin_number)) {
        if (i == 1) {
          bin_mean_result[[i]] <- rowMeans(matrix[, 
                                                  1:bin_size])
        }
        else {
          bin_mean_result[[i]] <- rowMeans(matrix[, 
                                                  (bin_size * (i - 1) + 1):(bin_size * i)])
        }
      }
      binned_matrix <- do.call(cbind, bin_mean_result)
      return(binned_matrix)
    }
    temp <- SummarizedExperiment(assays = lapply(assays(ChIPprofile_for_proplyr), 
                                                 bin_matrix, bin_size = bin_size), rowRanges = rowRanges(ChIPprofile_for_proplyr), 
                                 metadata = metadata(ChIPprofile_for_proplyr))
    ChIPprofile_for_proplyr = new("ChIPprofile", temp, 
                                  params = params(ChIPprofile_for_proplyr))
  }
}
all_profiles = ChIPprofile_for_proplyr 

assays <- lapply(1:length(bigwigfiles), function(i) assay(all_profiles, i))

# check if row names identical 
identical_row_names <- all(sapply(assays, function(x) identical(row.names(x), row.names(assays[[1]]))))

if(identical_row_names) {
  merged <- do.call(cbind, assays)
  rownames(merged) <-  rowData(all_profiles)$name
} else {
  print("Error: row names are not identical for all assays")
  merged <- NULL
}

saveRDS(merged, "fig7_atacseq_heatmap_parking.RDS")
```



```{r hm_clsrow_dend_reorder_f_row_annot, fig.width=6.5, fig.height=7, fig.path='fig7_atacseq_heatmap_parking/', dev=c('png'), eval=T}
all = readRDS("fig7_atacseq_heatmap_parking.RDS")
peaks = df
all = all[df$peak,]

# To improve visibility, bins with read counts greater than the 75th percentile + 1.5*IQR were capped at that value.
all[all >= as.numeric((quantile(all, c(0, .75))+IQR(all)*1.5)[2])] <- as.numeric((quantile(all, c(0, .75))+IQR(all)*1.5)[2])
colnames(all)  = seq(1:ncol(all))

names =   c("N.bw", "M.bw", "L1d.bw", "L5d.bw", "L10d.bw", "P1d.bw", "P5d.bw","P10d.bw")
annot_col = data.frame(cond =rep(gsub(".bw", "", names), each=100),row.names=colnames(all))
annot_col$cond = factor(annot_col$cond, levels=unique(rep(gsub(".bw", "", names), each=100)))

table(df$peak == row.names(all))

ha <- ComplexHeatmap::HeatmapAnnotation(
  foo = ComplexHeatmap::anno_block(
    gp = gpar(fill = c("#636363", "#00441b", "#439dcb", "#2171b5", "#2171b5", "#9e9ac8", "#807dba", "#6a51a3")),
    labels = unique(rep(gsub(".bw", "", names), each = 100)),
    labels_gp = gpar(col = rep("white", length(names)))
  )
)

hm = ComplexHeatmap::Heatmap(
  all,
  border = FALSE,
  cluster_rows = TRUE,
  cluster_columns = FALSE,
  col = rev(brewer.pal(n = 11, name = 'RdBu')),
  show_row_names = FALSE,
  heatmap_legend_param = list(direction = "horizontal", title = NULL, legend_width = unit(6, "cm")),
  column_split = annot_col$cond,
  top_annotation = ha,
  column_title = NULL,
  show_column_names = FALSE,
  use_raster = FALSE,
  show_row_dend = FALSE,
  row_split = df$km,
  cluster_row_slices = FALSE,
  row_title = NULL,
  row_gap = unit(0, "mm"),
  row_dend_reorder = FALSE
)
ComplexHeatmap::draw(hm, heatmap_legend_side = "bottom")
```

# Session Info
```{r session, message=FALSE, warning=FALSE, cache=FALSE,echo=FALSE, fig.width=10, fig.height=5.5, context="data"}
sessionInfo()
```