---
title: Ext. Fig. 4d
subtitle: "TF motif analysis comparing T cells activated in infected spleens vs tumor livers."
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output:
  html_document:
    code_folding: hide
    theme: space
    toc: yes
    toc_depth: 3
    toc_float: no
  BiocStyle::html_document2:
    code_folding: hide
    toc: yes
    toc_float: yes
  knitrBootstrap::bootstrap_document:
    highlight.chooser: yes
    theme.chooser: yes
  pdf_document:
    toc: yes
always_allow_html: yes
---


```{r setup, bootstrap.show.code = FALSE, results='hide', bootstrap.show.message=FALSE, warning=FALSE,  echo=FALSE, comment=FALSE}
knitr::opts_chunk$set(bootstrap.show.code = FALSE, message=FALSE, warning=FALSE)
suppressMessages(library(magrittr))
suppressMessages(library(kableExtra))
suppressMessages(library(ggplot2))
suppressMessages(library(data.table))
suppressMessages(library(ggrepel))
suppressMessages(library(DESeq2))
suppressMessages(library(DiffBind))
suppressMessages(library(dplyr))
suppressMessages(library(pheatmap))
suppressMessages(library(RColorBrewer))
suppressMessages(library(openxlsx))
suppressMessages(library(VennDiagram))
suppressMessages(library(chromVAR))
suppressMessages(library(chromVARmotifs))
suppressMessages(library(viridis))
suppressMessages(library(MotifDb))
suppressMessages(library(motifmatchr))
suppressMessages(library(BSgenome.Mmusculus.UCSC.mm10))
suppressMessages(library(motifmatchr))
set.seed(123)
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")
```


```{r read_in_atac_metadata, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE, context="data"}
decoderFile <- "atac_decoder_diffBind.csv"
decoder.data <- read.table(decoderFile,header=T,stringsAsFactors=F, sep=",")
decoder.data$Condition = factor(decoder.data$Condition)
decoder.data$Condition <- relevel(decoder.data$Condition, ref = "N")
```

```{r read_in_atac_counts, message=FALSE, warning=FALSE, cache=FALSE, echo=FALSE,  fig.width=10, fig.height=7, context="data"}
DB <- readRDS("DB.count.Rds")
DB <- dba.count(DB, peaks=NULL, score=DBA_SCORE_READS)
consensus_peaks <- dba.peakset(DB, peaks=NULL, bRetrieve=TRUE)
```

```{r createATACseqDDSobject, message=FALSE, warning=FALSE, cache=FALSE, echo=FALSE,  fig.width=10, fig.height=7, context="data"}
raw_counts = as.data.frame(mcols(consensus_peaks))
row.names(raw_counts) = paste0(as.data.frame(consensus_peaks)$seqnames, ":",as.data.frame(consensus_peaks)$start,"-",as.data.frame(consensus_peaks)$end)

#table(decoder.data$SampleID == colnames(raw_counts))

dds <- DESeqDataSetFromMatrix(countData = raw_counts,
                              colData = decoder.data,
                              design= ~ Condition)
dds <- DESeq(dds)
```

```{r read_in_rna_counts, message=FALSE, warning=FALSE,  echo=FALSE, context="data"}
counts <- read.table(file = "rnaseq.gene.counts.txt", header = TRUE, check.names=FALSE, row.names=1)
decoderFile <- "rnaseq.decoder.txt"
RNA.decoder.data <- fread(decoderFile) %>% as.data.frame()
RNA.decoder.data$group <- factor(RNA.decoder.data$group, levels=c("N", "E6", "E12", "E24","L6", "L12", "L24"))
RNA.decoder.data$type <- factor(RNA.decoder.data$type, levels=c("naive", "effector", "tumor"))
RNA.decoder.data$replicate <- factor(RNA.decoder.data$replicate)
RNA.decoder.data$time <- factor(RNA.decoder.data$time, levels=c("0hr", "6hr", "12hr", "24hr"))

# keep 6 and 12 hr
RNA.decoder.data = subset(RNA.decoder.data, time == "6hr" | time == "12hr")

RNA.decoder.data <- RNA.decoder.data[RNA.decoder.data$sample.ID %in% colnames(counts),]
counts <- counts[,c(RNA.decoder.data$sample.ID)]

# compute cpm 
cpms = edgeR::cpm(counts)
# remove genes that don't have baseMean > 10 in each rowMean
cpms_filt = cpms[rowMeans(edgeR::cpm(counts)) > 10,]

#colnames(counts) == RNA.decoder.data$sample.ID 
```

# chromVar 

Running chromVar on motifs with expressed TFs rowMeans(cpm) > 10 on E6/12, L6/12 only

```{r summarizeChromVARMotifs,  eval=T,  warning=FALSE,  echo=FALSE,  fig.width=5, fig.height=5}

# summarizeChromVARMotifs function from archR: https://github.com/GreenleafLab/ArchR/blob/f6c0388bd37023400794c9ae8562ad69e3ba9fd7/R/AnnotationPeaks.R

summarizeChromVARMotifs <- function(motifs = NULL){

  motifNames <- lapply(seq_along(motifs), function(x){
    namex <- make.names(motifs[[x]]@name)
    if(grepl("LINE", namex)){
      splitNamex <- stringr::str_split(motifs[[x]]@ID, pattern="\\_", simplify = TRUE)
      namex <- splitNamex[1, grep("LINE",splitNamex[1,]) + 1]
    }
    if(substr(namex,nchar(namex),nchar(namex))=="."){
      namex <- substr(namex,1,nchar(namex)-1)
    }
    namex <- paste0(namex, "_", x)
    namex
  }) %>% unlist(.)

  motifNames2 <- lapply(seq_along(motifs), function(x){
    namex <- make.names(motifs[[x]]@name)
    if(grepl("LINE", namex)){
      splitNamex <- stringr::str_split(motifs[[x]]@ID, pattern="\\_", simplify = TRUE)
      namex <- splitNamex[1, grep("LINE",splitNamex[1,]) + 1]
    }
    if(substr(namex,nchar(namex),nchar(namex))=="."){
      namex <- substr(namex,1,nchar(namex)-1)
    }
    namex
  }) %>% unlist(.)

  motifDF <- lapply(seq_along(motifs), function(x){
    df <- data.frame(
      row.names = motifNames[x],
      name = motifNames2[[x]],
      ID = motifs[[x]]@ID,
      strand = motifs[[x]]@strand,
      stringsAsFactors = FALSE
    )
  }) %>% Reduce("rbind", .) %>% DataFrame

  names(motifs) <- motifNames

  out <- list(motifs = motifs, motifSummary = motifDF)

  return(out)

}
```

```{r chromVar,  eval=T,  warning=FALSE,  echo=FALSE,  fig.width=5, fig.height=5}
set.seed(123)

motifs_expr <- mouse_pwms_v2
motif_names = summarizeChromVARMotifs(motifs_expr)$motifSummary %>% as.data.frame()

# only retain motifs that are expressed in the RNA-seq data
motif_names = motif_names[motif_names$name %in% row.names(cpms_filt),]
motifs_expr = motifs_expr[which(names(motifs_expr) %in% motif_names$ID)]


motifs <- motifs_expr

decoder.data.sub = subset(decoder.data, time == "6hr" | time == "12hr")
kable(decoder.data.sub, row.names=FALSE,  padding = 0, longtable=TRUE) %>%  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = F)

# https://support.bioconductor.org/p/71301/
DB_se = dba(DB, bSummarizedExperiment=TRUE)
DB_se = DB_se[,gsub("\\.", "-", decoder.data.sub$SampleID)]
assayNames(DB_se)[3] <- "counts"

counts <- DB_se[rowSums(assay(DB_se)) > 5, ]
counts <- addGCBias(counts, genome = BSgenome.Mmusculus.UCSC.mm10)

# The function matchMotifs from the motifmatchr package finds which peaks contain which motifs. 
motif_ix <- matchMotifs(motifs, rowRanges(counts), genome = BSgenome.Mmusculus.UCSC.mm10)
dev <- computeDeviations(object = counts, annotations = motif_ix)
variability <- computeVariability(dev)


res_df <- cbind(variability, rank = rank(-1 * variability$variability, ties.method = "random"))
res_df <- res_df[order(res_df$rank, decreasing = F), ]
res_df$rank <- NULL
res_df = res_df[res_df$p_value_adj < 0.05,]


n=25
topVariable <- res_df[1:n, ]
if(length(grep("LINE", topVariable$name)) > 0){
  topVariable[grep("LINE", topVariable$name),]$name = sapply(strsplit(row.names(topVariable[grep("LINE", topVariable$name),]), "_"), function(x) x[4])
}
devTop <- merge(topVariable[, 1, drop = FALSE], assay(dev, "z"), by = 0)
devToPlot <- as.matrix(devTop[, -c(1:2)])
rownames(devToPlot) <- devTop[, 2]


devToPlot_ordered = devToPlot
col_order = gtools::mixedsort(colnames(devToPlot_ordered))
devToPlot_ordered = devToPlot_ordered[,col_order]

annot_col = data.frame( colData(dev))[,"Condition", drop=F]

annotation_colors = list(Condition=c("E6"="#ccece6",
                                     "E12"="#99d8c9",
                                     "L6" = "#c6dbef",
                                     "L12" = "#6baed6"))

pheatmap(devToPlot_ordered, 
         scale="none", 
         annotation_col =annot_col,
         annotation_colors = annotation_colors,
         cluster_cols = F,
         cluster_rows=T,
         show_rownames=T,
         clustering_method = "ward.D2",
        color = viridis(50), 
        fontsize_row = 12, 
        gaps_col = 6,
        silent = F, 
        cellwidth=15, 
        cellheight=10,  
        treeheight_col=45,
        main = "", 
        border_color=NA,
        treeheight_row=0)

```




# Session Info
```{r session, message=FALSE, warning=FALSE, cache=FALSE,echo=FALSE, fig.width=10, fig.height=5.5, context="data"}
sessionInfo()
```
