---
title: Fig. 4e
subtitle: "Combined ATAC-SEQ and RNA-SEQ analysis of TCRTAG T cells 6 hours after activation in tumor livers relative to infected spleens showing top 25 DAC with corresponding differentially expressed genes (DEG)"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output:
  html_document:
    code_folding: hide
    theme: space
    toc: yes
    toc_depth: 3
    toc_float: no
  BiocStyle::html_document2:
    code_folding: hide
    toc: yes
    toc_float: yes
  knitrBootstrap::bootstrap_document:
    highlight.chooser: yes
    theme.chooser: yes
  pdf_document:
    toc: yes
always_allow_html: yes
---

```{r setup, bootstrap.show.code = FALSE, results='hide', bootstrap.show.message=FALSE, warning=FALSE,  echo=FALSE}
knitr::opts_chunk$set(bootstrap.show.code = FALSE, message=FALSE, warning=FALSE)
suppressMessages(library(magrittr))
suppressMessages(library(kableExtra))
suppressMessages(library(ggplot2))
suppressMessages(library(data.table))
suppressMessages(library(ggrepel))
suppressMessages(library(DESeq2))
suppressMessages(library(DiffBind))
suppressMessages(library(dplyr))
suppressMessages(library(pheatmap))
suppressMessages(library(RColorBrewer))
suppressMessages(library(VennDiagram))
suppressMessages(library(TxDb.Mmusculus.UCSC.mm10.knownGene))
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")
```



```{r read_in_atac_metadata, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE, context="data"}
decoderFile <- "atac_decoder_diffBind.csv"
decoder.data <- read.table(decoderFile,header=T,stringsAsFactors=F, sep=",")
decoder.data$Condition = factor(decoder.data$Condition)
decoder.data$Condition <- relevel(decoder.data$Condition, ref = "N")
```

```{r read_in_atac_counts, message=FALSE, warning=FALSE, cache=FALSE, echo=FALSE,  fig.width=10, fig.height=7, context="data"}
DB <- readRDS("DB.count.Rds")
DB <- dba.count(DB, peaks=NULL, score=DBA_SCORE_READS)
consensus_peaks <- dba.peakset(DB, peaks=NULL, bRetrieve=TRUE)
consensus_peaks_annot = ChIPseeker::annotatePeak(consensus_peaks,tssRegion = c(-1000,1000),TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,annoDb = "org.Mm.eg.db", verbose=F)
consensus_peaks_annot_df = as.data.frame(consensus_peaks_annot)
row.names(consensus_peaks_annot_df) =  paste0(as.data.frame(consensus_peaks_annot_df)$seqnames, ":",as.data.frame(consensus_peaks_annot_df)$start,"-",as.data.frame(consensus_peaks_annot_df)$end)
```


# Differential accessibility (ATAC-seq)

The following peaks were detected as differentially accessible:


```{r createATACseqDDSobject, message=FALSE, warning=FALSE,  echo=FALSE,  fig.width=10, fig.height=7, context="data"}
raw_counts = as.data.frame(mcols(consensus_peaks))
row.names(raw_counts) = paste0(as.data.frame(consensus_peaks)$seqnames, ":",as.data.frame(consensus_peaks)$start,"-",as.data.frame(consensus_peaks)$end)

#table(decoder.data$SampleID == colnames(raw_counts))

dds <- DESeqDataSetFromMatrix(countData = raw_counts,
                              colData = decoder.data,
                              design= ~ Condition)
dds <- DESeq(dds, fitType = "parametric")
```

```{r da_peaks, eval=T,echo=FALSE, message=FALSE, warning=FALSE,   context="data"}
varInt="Condition"
dba.all.peaks <- list()
alphaTR = 0.05
lfcThres = 1

for (comp in combn(nlevels(colData(dds)[,varInt]), 2, simplify=FALSE)){
  levelRef <- levels(colData(dds)[,varInt])[comp[1]]
  levelTest <- levels(colData(dds)[,varInt])[comp[2]]
  cat(paste("Comparison", levelTest, "vs", levelRef, "done\n"))
  dba.all.peaks[[paste0(levelTest,"_vs_",levelRef)]] <- 
    results(dds, contrast=c(varInt, levelTest, levelRef), alpha=alphaTR) %>% as.data.frame()
}


dba.all.peaks.sig <- lapply(names(dba.all.peaks), function(x){
  subset(dba.all.peaks[[x]], padj < alphaTR & abs(log2FoldChange) > lfcThres)
   })
names(dba.all.peaks.sig) <- names(dba.all.peaks)
 
 
res = data.frame(sapply(dba.all.peaks.sig, NROW))
colnames(res) <- paste0("padj<",alphaTR, " & ","abs(log2FoldChange)>",lfcThres)


all.comps = stringr::str_split(row.names(res), "_vs_")
rel.contrs = 1:length(all.comps)

kable(res, row.names=T,  padding = 0, longtable=TRUE) %>%  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = F)
```


# Differential expression (RNA-seq)

```{r read_in_rna_counts, message=FALSE, warning=FALSE,  echo=FALSE, context="data"}
counts <- read.table(file = "rnaseq.gene.counts.txt", header = TRUE, check.names=FALSE, row.names=1)
decoderFile <- "rnaseq.decoder.txt"
decoder.data <- fread(decoderFile) %>% as.data.frame()
decoder.data$group <- factor(decoder.data$group, levels=c("N", "E6", "E12", "E24","L6", "L12", "L24"))
decoder.data$type <- factor(decoder.data$type, levels=c("naive", "effector", "tumor"))
decoder.data$replicate <- factor(decoder.data$replicate)
decoder.data$time <- factor(decoder.data$time, levels=c("0hr", "6hr", "12hr", "24hr"))

decoder.data <- decoder.data[decoder.data$sample.ID %in% colnames(counts),]
counts <- counts[,c(decoder.data$sample.ID)]

#colnames(counts) == decoder.data$sample.ID 
```

```{r createRNAseqDDSobject, message=FALSE, warning=FALSE,  echo=FALSE,  fig.width=10, fig.height=7, context="data"}
rna_dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = decoder.data,
                              design= ~ group)
rna_dds <- DESeq(rna_dds)
```

The following genes were detected as differentially expressed:

```{r rna_de, eval=T,echo=FALSE, message=FALSE, warning=FALSE,   context="data"}
varInt="group"
rna_dbs.all <- list()
alphaTR = 0.05
lfcThres = 1

rna_dbs.all <- list()
for (comp in combn(nlevels(colData(rna_dds)[,varInt]), 2, simplify=FALSE)){
  levelRef <- levels(colData(rna_dds)[,varInt])[comp[1]]
  levelTest <- levels(colData(rna_dds)[,varInt])[comp[2]]
  cat(paste("Comparison", levelTest, "vs", levelRef, "done\n"))
  rna_dbs.all[[paste0(levelTest,"_vs_",levelRef)]] <- 
    results(rna_dds, contrast=c(varInt, levelTest, levelRef), alpha=alphaTR) %>% as.data.frame()
}


rna_dbs.all.sig <- lapply(names(rna_dbs.all), function(x){
  subset(rna_dbs.all[[x]], padj < alphaTR  & abs(log2FoldChange) > lfcThres)
   })
names(rna_dbs.all.sig) <- names(rna_dbs.all)
 
 
res = data.frame(sapply(rna_dbs.all.sig, NROW))
colnames(res) <-  paste0("padj<",alphaTR, " & ","abs(log2FoldChange)>",lfcThres)

kable(res, row.names=T,  padding = 0, longtable=TRUE) %>%  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = F)
```


# Diamond plots (peaks) {.tabset}

For diamond plots, selected genes that were DE (FDR < 0.05 & |log2FC| > 1 & baseMean > 10 ) with at least one (non-intergenic) peak that was DA (FDR < 0.05). Then plotted the top / bottom 25 genes ordered by the DE stat column.

## L6_vs_E6

```{r diamond_L6_vs_E6_top25, eval=T,  warning=FALSE, echo=FALSE,  fig.width=5, fig.height=8, results="asis"}
dia_rnaseq =  rna_dbs.all.sig$L6_vs_E6[rna_dbs.all.sig$L6_vs_E6$baseMean > 10,]
dia_rnaseq = dia_rnaseq[order(dia_rnaseq$stat),]
dia_rnaseq$SYMBOL = row.names(dia_rnaseq)
dia_rnaseq$Row.names = row.names(dia_rnaseq)
dia_rnaseq$Color <- ifelse(dia_rnaseq$padj < 0.05 & dia_rnaseq$log2FoldChange < 0,  "#67a9cf", ifelse(dia_rnaseq$padj < 0.05 & dia_rnaseq$log2FoldChange >0, "#ef8a62", "darkgray"))

dia_ATAC = cbind(dba.all.peaks$L6_vs_E6, consensus_peaks_annot_df)
dia_ATAC$Color <- ifelse(!is.na(dia_ATAC$padj) & dia_ATAC$padj < 0.05 & dia_ATAC$log2FoldChange < 0,  "#67a9cf", ifelse(!is.na(dia_ATAC$padj) & dia_ATAC$padj < 0.05 & dia_ATAC$log2FoldChange >0, "#ef8a62", "darkgray"))
dia_ATAC = subset(dia_ATAC, annotation != "Distal Intergenic")

inter = unique(intersect(dia_rnaseq$SYMBOL, subset(dia_ATAC, padj < 0.05)$SYMBOL))

up_in_L6 = subset(dia_rnaseq[inter,], log2FoldChange > 0)
up_in_E6 = subset(dia_rnaseq[inter,], log2FoldChange < 0)
#write.table(row.names(up_in_L6), "up_in_L6.forEnrichr.txt", quote = F, row.names=F, col.names=F)
#write.table(row.names(up_in_E6), "up_in_E6.forEnrichr.txt", quote = F, row.names=F, col.names=F)

dia_rnaseq = subset(dia_rnaseq, SYMBOL %in% inter)
dia_ATAC = subset(dia_ATAC, SYMBOL %in% inter)

goi = psych::headTail(dia_rnaseq, top=25, bottom=25, ellipsis = F)$Row.names
dia_rnaseq = subset(dia_rnaseq, Row.names %in% goi)
dia_ATAC_sub = subset(dia_ATAC, SYMBOL %in% goi)

dia_rnaseq$SYMBOL = factor(dia_rnaseq$SYMBOL, levels=dia_rnaseq[order(dia_rnaseq$log2FoldChange),]$SYMBOL)
dia_ATAC_sub$SYMBOL = factor(dia_ATAC_sub$SYMBOL, levels=dia_rnaseq[order(dia_rnaseq$log2FoldChange),]$SYMBOL)

p1 = ggplot(dia_ATAC_sub, aes(x=log2FoldChange, y=SYMBOL, color=Color)) + geom_point(shape=23, size=3) + theme_bw() + scale_color_identity() + geom_vline(xintercept=c(0), linetype="dotted") + geom_hline(yintercept=min(which(dia_rnaseq$log2FoldChange > 0)), linetype="dotted") + xlab("peak accessibility \n log2(T6h v E6h)") + ylab("")  +theme(axis.text=element_text(size=12))

p2 = ggplot(dia_rnaseq, aes(x=log2FoldChange, y=SYMBOL, fill=Color)) + geom_point(shape=21, size=4) + theme_bw() + scale_fill_identity() + geom_vline(xintercept=c(0), linetype="dotted") + geom_hline(yintercept=min(which(dia_rnaseq$log2FoldChange > 0)), linetype="dotted") + xlab("gene expression\n log2(T6h v E6h)") + theme(axis.title.y=element_blank(),  axis.text.y=element_blank()) 

p1 + p2 
```


# Session Info
```{r session, message=FALSE, warning=FALSE, cache=FALSE,echo=FALSE, fig.width=10, fig.height=5.5, context="data"}
sessionInfo()
```