---
title: Fig. 4f
subtitle: "Heatmap of differentially expressed genes DEG from bulk RNA-SEQ"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output:
  html_document:
    code_folding: hide
    theme: space
    toc: yes
    toc_depth: 3
    toc_float: no
  BiocStyle::html_document2:
    code_folding: hide
    toc: yes
    toc_float: yes
  knitrBootstrap::bootstrap_document:
    highlight.chooser: yes
    theme.chooser: yes
  pdf_document:
    toc: yes
always_allow_html: yes
---

```{r setup, bootstrap.show.code = FALSE, results='hide', bootstrap.show.message=FALSE, warning=FALSE,  echo=FALSE}
knitr::opts_chunk$set(bootstrap.show.code = FALSE, message=FALSE, warning=FALSE)
suppressMessages(library(magrittr))
suppressMessages(library(kableExtra))
suppressMessages(library(ggplot2))
suppressMessages(library(data.table))
suppressMessages(library(ggrepel))
suppressMessages(library(DESeq2))
suppressMessages(library(DiffBind))
suppressMessages(library(dplyr))
suppressMessages(library(pheatmap))
suppressMessages(library(RColorBrewer))
suppressMessages(library(VennDiagram))
suppressMessages(library(TxDb.Mmusculus.UCSC.mm10.knownGene))
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")
```



# Samples

```{r read_in_rna_counts, message=FALSE, warning=FALSE,  echo=FALSE, context="data"}
counts <- read.table(file = "rnaseq.gene.counts.txt", header = TRUE, check.names=FALSE, row.names=1)
decoderFile <- "rnaseq.decoder.txt"
decoder.data <- fread(decoderFile) %>% as.data.frame()
decoder.data$group <- factor(decoder.data$group, levels=c("N", "E6", "E12", "E24","L6", "L12", "L24"))
decoder.data$type <- factor(decoder.data$type, levels=c("naive", "effector", "tumor"))
decoder.data$replicate <- factor(decoder.data$replicate)
decoder.data$time <- factor(decoder.data$time, levels=c("0hr", "6hr", "12hr", "24hr"))

decoder.data <- decoder.data[decoder.data$sample.ID %in% colnames(counts),]
counts <- counts[,c(decoder.data$sample.ID)]

kable(decoder.data, row.names=FALSE,  padding = 0, longtable=TRUE) %>%  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
#colnames(counts) == decoder.data$sample.ID 
```

```{r createRNAseq_ddsobject, message=FALSE, warning=FALSE,  echo=FALSE,  fig.width=10, fig.height=7, context="data"}
rna_dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = decoder.data,
                              design= ~ group)
rna_dds <- DESeq(rna_dds)
rna_vst = vst(rna_dds)
normCounts = assay(rna_vst)
```

# Differential expression (RNA-seq)

The following genes were detected as differentially expressed:

```{r rna_de, eval=T,echo=FALSE, message=FALSE, warning=FALSE,   context="data"}
varInt="group"
rna_dbs.all <- list()
alphaTR = 0.05
lfcThres = 1

rna_dbs.all <- list()
for (comp in combn(nlevels(colData(rna_dds)[,varInt]), 2, simplify=FALSE)){
  levelRef <- levels(colData(rna_dds)[,varInt])[comp[1]]
  levelTest <- levels(colData(rna_dds)[,varInt])[comp[2]]
  cat(paste("Comparison", levelTest, "vs", levelRef, "done\n"))
  rna_dbs.all[[paste0(levelTest,"_vs_",levelRef)]] <- 
    results(rna_dds, contrast=c(varInt, levelTest, levelRef), alpha=alphaTR) %>% as.data.frame()
}


rna_dbs.all.sig <- lapply(names(rna_dbs.all), function(x){
  subset(rna_dbs.all[[x]], padj < alphaTR  & abs(log2FoldChange) > lfcThres)
   })
names(rna_dbs.all.sig) <- names(rna_dbs.all)
 
 
res = data.frame(sapply(rna_dbs.all.sig, NROW))
colnames(res) <-  paste0("padj<",alphaTR, " & ","abs(log2FoldChange)>",lfcThres)

kable(res, row.names=T,  padding = 0, longtable=TRUE) %>%  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = F)
```


# Heatmap of select genes

```{r heatmap_goi, eval=T,echo=FALSE, message=FALSE, warning=FALSE,   context="data"}
negative_regulators = c("Rgs16", "Dusp4", "Pdcd1", "Cd226", "Ptpn22", "Ctla4", "Tigit")
inflammatory_genes = c("Mx1", "Ifit3", "Oas2", "Isg15")
effector_genes = c("Ifng", "Gzmb", "Gzma", "Fasl")
tfs = c("Irf7", "Batf", "Eomes", "Stat1", "Tbx21", "Bach2", "Id3")
all = c(negative_regulators,inflammatory_genes,effector_genes,tfs)

breaksList = seq(-1.5, 1.5, by = 0.1)

decoder.sub = subset(decoder.data, time == "12hr" | time == "6hr")
row.names(decoder.sub) = decoder.sub$sample.ID
decoder.sub = decoder.sub[order(decoder.sub$group),]
pheatmap(normCounts[negative_regulators,decoder.sub$sample.ID],scale="row", color = colorRampPalette(rev(brewer.pal(n = 5, name = "RdBu")))(length(breaksList)), breaks = breaksList, cluster_rows = F, cluster_cols = F, annotation_col = decoder.sub[,"group", drop=F], gaps_col = 6, cellwidth = 25, cellheight = 25, main = "Negative Regulators")


pheatmap(normCounts[inflammatory_genes,decoder.sub$sample.ID],scale="row", color = colorRampPalette(rev(brewer.pal(n = 5, name = "RdBu")))(length(breaksList)), breaks = breaksList, cluster_rows = F, cluster_cols = F, annotation_col = decoder.sub[,"group", drop=F], gaps_col = 6, cellwidth = 25, cellheight = 25, main = "Inflammatory Genes")


pheatmap(normCounts[effector_genes,decoder.sub$sample.ID],scale="row", color = colorRampPalette(rev(brewer.pal(n = 5, name = "RdBu")))(length(breaksList)), breaks = breaksList, cluster_rows = F, cluster_cols = F, annotation_col = decoder.sub[,"group", drop=F], gaps_col = 6, cellwidth = 25, cellheight = 25, main = "Effector Genes")


pheatmap(normCounts[tfs,decoder.sub$sample.ID],scale="row", color = colorRampPalette(rev(brewer.pal(n = 5, name = "RdBu")))(length(breaksList)), breaks = breaksList, cluster_rows = F, cluster_cols = F, annotation_col = decoder.sub[,"group", drop=F], gaps_col = 6, cellwidth = 25, cellheight = 25, main = "Transcription Factors")


```


All of the above genes are DE in in either L6_vs_E6 or L12_vs_E12 (`r all[(all %in% c(row.names(rna_dbs.all.sig$L12_vs_E12), row.names(rna_dbs.all.sig$L6_vs_E6)))]`), except `r all[!(all %in% c(row.names(rna_dbs.all.sig$L12_vs_E12), row.names(rna_dbs.all.sig$L6_vs_E6)))]`

# Session Info
```{r session, message=FALSE, warning=FALSE, cache=FALSE,echo=FALSE, fig.width=10, fig.height=5.5, context="data"}
sessionInfo()
```